ambry-api/src/main/java/com/github/ambry/quota/QuotaManager.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota;
+
+import com.github.ambry.messageformat.BlobInfo;
+import com.github.ambry.rest.RestRequest;
+import java.util.List;
+
+
+/**
+ * Interface for the class that acts as the manager of all the quotas in Ambry. Implementations of this interface
+ * should take care of initializing all the various type of quota enforcements and getting the overall quota
+ * recommendation for each request.
+ */
+public interface QuotaManager {
+
+  /**
+   * Method to initialize the {@link QuotaManager}.
+   */
+  void init();
+
+  /**
+   * Computes the overall boolean recommendation to throttle a request or not, for all the types of quotas that depend on
+   * the load on host resources, and populates the specified enforcementRecommendations list with the
+   * recommendations. This method should be used for throttling on quota that doesn't depend upon request characteristics.
+   * @param enforcementRecommendations {@link List} of {@link EnforcementRecommendation}s to be populated.
+   * @return true if the request should be throttled. false otherwise.
+   */
+  boolean shouldThrottleOnHost(List<EnforcementRecommendation> enforcementRecommendations);
+
+  /**
+   * Computes the overall boolean recommendation to throttle a request or not for all the types of request quotas supported.
```

Comment:
There is will be only one instance of QuotaManager class in the process (like say ReplicationManager). On its instantiation, QuotaManager will create HostQuotaEnforcer and  RequestQuotaEnforcer classes based on config values in QuotaConfig. 

On each request, wherever there is a need to do quota check or enforcement, QuotaManager.shouldThrottle* methods will be called. These methods will ask for recommendation from each of the enforcers, and will forward the recommendation to the caller. The caller can now decide what it wants to do with the recommendation. 

For now, the only thing caller does with the recommendation is an AND of all the quotaThrottle flags of all the recommendations. But we might need richer recommendation for future. For example, on each request, we might also want to send out the request cost, and usage percentage to the users as http response headers. Also going forward, we might want to add the concept of something like QuotaPolicy (not implemented yet). Using quota policy we can specify more complex throttling mechanisms like maybe throttle only for CAPACITY_UNIT and not storage, or throttle for certain containers while dont throttle for certain other high priority containers etc.

Due to all these reasons we have EnforcementRecommendation being returned. And since there is only one global QuotaManager instance, the recommendation list needs to be initialized for each request. Hence we cannot have then pre instantitated.